src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcEndpointDataSource.cs

```diff
@@ -392,36 +397,36 @@ private IChangeToken GetCompositeChangeToken()
             return new CompositeChangeToken(changeTokens);
         }
 
-        public override IChangeToken ChangeToken
-        {
-            get
-            {
-                if (_changeToken == null)
-                {
-                    _changeToken = GetCompositeChangeToken();
-                }
-
-                return _changeToken;
-            }
-        }
+        public override IChangeToken ChangeToken => GetCompositeChangeToken();
 
         public override IReadOnlyList<Endpoint> Endpoints
         {
             get
             {
-                if (!_initialized)
+                // Want to initialize endpoints once and then cache while ensuring a null collection is never returned
+                // Local copy for thread safety + double check locking
+                var localEndpoints = _endpoints;
+                if (localEndpoints == null)
                 {
                     lock (_lock)
                     {
-                        if (!_initialized)
+                        localEndpoints = _endpoints;
+                        if (localEndpoints == null)
                         {
-                            InitializeEndpoints();
-                            _initialized = true;
+                            _endpoints = localEndpoints = CreateEndpoints();
                         }
                     }
                 }
 
-                return _endpoints;
+                return localEndpoints;
```

Comment:
I'd suggest just using `LazyInitializer.EnsureInitialized` for this.

Reply:
I'll leave it as is for now. We can re-evaluate in the future if it changes